### ISSUE TITLE
Issue executing Magick.NET from lambda

### ISSUE THREAD

**Issue by ytn3rd** [aws/aws-lambda-dotnet#153]
I've been trying to do some ImageMagick work using dotnet in lambda. I was previously doing it with node.js which would execute a pre-compiled php binary and we'd do the work in php. But that seems pretty excessive.

I had initially thought the issue was with the Magick.NET library itself, but I posted over in their [github issues section](https://github.com/dlemstra/Magick.NET/issues/94) (including some sample code) about the problems I was having, mainly this:

> {
>   "errorType": "TypeInitializationException",
>   "errorMessage": "The type initializer for 'NativeMagickSettings' threw an exception.",
>   "stackTrace": [
>     "at ImageMagick.MagickSettings.NativeMagickSettings..ctor()",
>     "at ImageMagick.MagickSettings..ctor()",
>     "at ImageMagick.MagickImage..ctor()",
>     "at LambdaImageMagickTest.Program.MyHandler(ILambdaContext context) in /Users/bradmoore/Bitbucket/videomyjob-lambda/LambdaImageMagickTest/Program.cs:line 153",
>     "at lambda_method(Closure , Stream , Stream , ContextInfo )"
>   ],
>   "cause":   {
>     "errorType": "DllNotFoundException",
>     "errorMessage": "Unable to load DLL 'Magick.NET-Q8-x64.Native.dll': The specified module could not be found.
>  (Exception from HRESULT: 0x8007007E)",
>     "stackTrace": [
>       "at ImageMagick.Environment.NativeMethods.X64.Environment_Initialize()",
>       "at ImageMagick.Environment.Initialize()",
>       "at ImageMagick.MagickSettings.NativeMagickSettings..cctor()"
>     ]
>   }
> }

They reference this error over in their [crossplatform](https://github.com/dlemstra/Magick.NET/blob/master/Documentation/CrossPlatform.md) doc and I followed the instructions and got all the .so files and put them in my projects lib directory. ldd now reports nothing missing executing all those .so files. The author of Magick.NET asked if I can use strace to get a stacktrace of what is missing when we try to execute the project. After getting the strace binary in my project (via another Amazon Linux server) but apparently strace isn't supported on the other platform (maybe some security?)

> test_ptrace_setoptions_for_all: PTRACE_TRACEME doesn't work: Operation not permitted\n/var/task/strace: test_ptrace_setoptions_for_all: unexpected exit status 1

While attempting to get strace running I had to run my project via a new Process, dotnet run blah.dll, but dotnet was not found (dotnet isn't installed on the platform executing my dotnet code?). So I included dotnet-centos-x64.1.0.5.tar.gz in my project, extracted it into tmp and executed my code using this dotnet version and my Magick.NET code executed. 

The code itself is just,  

```
var magickImage = new MagickImage(MagickColors.Red, 100, 100);
output += $"Image size: {magickImage.Width}x{magickImage.Height}";
```

It is not trying to do anything special (yet :P)

So I am back at square one. 

Any idea why my code would execute in lambda only via a new dotnet install and would fail on the dotnet install the system is currently running?

I've tried to find out what version of dotnet runtime is executing my code but internet say LTS version. Is it using 1.0.5 as that is the current LTS version, is it still on 1.0.4? Is that what is causing the issue or is something else at play here?

EDIT: I did also publish my code with dotnet lambda package and it didn't report any issues with any references to any .NET Standard libraries I shouldn't be using that other people have reported.

**Comment by normj** [comment]
I haven't tried using ImageMagick before. Any chance you could put together a small GitHub repo with the code reproduce the problem?

**Comment by ytn3rd** [comment]
I have attached the sample code from the other post. Its enough to see that it dies. You will need to uncomment the `var magickImage = new ...` line and the one below it. 

Apparently GitHub is having issues with me uploading the source directly into a repo this morning. 

[LambdaImageMagickTest.publish.zip](https://github.com/aws/aws-lambda-dotnet/files/1268808/LambdaImageMagickTest.publish.zip)
[LambdaImageMagickTest.source.zip](https://github.com/aws/aws-lambda-dotnet/files/1268807/LambdaImageMagickTest.source.zip)

**Comment by ytn3rd** [comment]
@normj , unsure what was up with github the other day, but I have pushed another sample to a test git repo.

[https://github.com/ytn3rd/lamba-dotnet-magick.net](https://github.com/ytn3rd/lamba-dotnet-magick.net)

App should work out of the box, but if you uncomment [these two](https://github.com/ytn3rd/lamba-dotnet-magick.net/blob/master/LambdaImageMagickTest/Program.cs#L50-L51) lines it will fail.

**Comment by ytn3rd** [comment]
@normj any chance you'd have time to check this as to why it is broken but works if I drag in my own dotnet runtimes into lambda and execute it via that?

**Comment by derwasp** [comment]
@ytn3rd just FYI, if you are talking about `Albelli.Magick.Builder`, please be aware that I don't have support for all the libraries there.

**Comment by jerhewet** [comment]
Just wanted to add some possible caveats to this thread, in case they might help anyone else.

I'm using the latest version of Magick.NET-Q16-AnyCPU, and the requisite DLL and SO files aren't copied into your build output directories (/bin/*) unless you publish the lambda.  I used the example that @beeradmoore published to his GitHub account, and everything is working perfectly for me now that I've published, both locally (Mock Lambda Test Tool) and on my AWS Lambda as well.

I'm still too new to AWS Lambda's and the AWS Visual Studio 2019 solution and project configurations to know if I'm missing anything obvious, but I was able to get everything to work, thanks to @beeradmoore taking the time to post his lambda solution to GitHub.
